### DISCUSSION TITLE
ValueError: cmap format 4 subtable overflowed; sort glyph order by unicode to fix.

### DISCUSSION THREAD

**Discussion by TakWolf** [fonttools/fonttools#3230] [Q&A]
```
https://github.com/fonttools/fonttools/blob/7fce3444046485846118c1eda86d4ed0a541a94c/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L1029

    header = struct.pack(
             ^^^^^^^^^^^^
struct.error: 'H' format requires 0 <= number <= 65535

During handling of the above exception, another exception occurred:

https://github.com/fonttools/fonttools/blob/7fce3444046485846118c1eda86d4ed0a541a94c/Lib/fontTools/fontBuilder.py#L415

ValueError: cmap format 4 subtable overflowed; sort glyph order by unicode to fix.
```

What does this mean? What should I sort? Should I sort `glyphOrder: list[str]` as asc `code_points`? Why this need a sequence?

I found an article：https://zhuanlan.zhihu.com/p/20100737
The point is: https://learn.microsoft.com/zh-cn/typography/opentype/spec/cmap#windows-platform-platform-id--3

```
Fonts that support only Unicode BMP characters (U+0000 to U+FFFF) on the Windows platform must use encoding 1 with a format 4 subtable. This encoding must not be used to support Unicode supplementary-plane characters.

Fonts that support Unicode supplementary-plane characters (U+10000 to U+10FFFF) on the Windows platform must use encoding 10 with a format 12 subtable.
```
Is it related to this？

Because if I set this: 

```    
builder.setupCharacterMap(character_mapping, allowFallback=True)
```
Compile without error, but the fonts will not support windows: (Not a valid font file)

![b20746e96b91373e22a1be84ecc3ecc1](https://github.com/fonttools/fonttools/assets/6064962/e257f49e-b4f1-43f3-a726-8cc6e87787e0)

By the way, in my case, all cmap code_point are < 0xFFFF. I don't understand the essence of the problem right now.

**Comment by justvanrossum** [comment]
Format 4 cmap uses some sort of segmented format, which works best if the glyph order is somewhat sorted by unicode. It will allow to compactly store runs of sequential code points that map to sequential glyph IDs. 

Unfortunately, there are cases with large cmaps and "unsorted" glyph IDs where this principle apparently degenerates (I'm not even sure of the details) and an overflow occurs. Having more of the glyph order sorted by unicode generally helps.

**Reply by justvanrossum** [reply]
Btw. the overflow does not mean there's a bug in fonttools but rather that this particular cmap/glyph order combination *can not be represented* with a format 4 cmap.

**Reply by TakWolf** [reply]
If I sort glyphOrder as code_point sequence, how many glyphs can be stored in theory? Is 65535 (Because 0x0000 ~ 0xFFFF)?

Now I have 43678 glyphs without sort, It is overflowed.

**Comment by behdad** [comment]
Does this patch help?
```diff
diff --git a/Lib/fontTools/ttLib/tables/_c_m_a_p.py b/Lib/fontTools/ttLib/tables/_c_m_a_p.py
index 6c00aaf63..6ecdf3807 100644
--- a/Lib/fontTools/ttLib/tables/_c_m_a_p.py
+++ b/Lib/fontTools/ttLib/tables/_c_m_a_p.py
@@ -1029,7 +1029,7 @@ class cmap_format_4(CmapSubtable):
         header = struct.pack(
             cmap_format_4_format,
             self.format,
-            length,
+            min(length, 0xFFFF),
             self.language,
             segCountX2,
             searchRange,
```

**Reply by TakWolf** [reply]
Amazing, it's effective!

I still haven't understood the principle. Does this mean that sorting is not necessary? (When conditions: all code points < 0xFFFF and glyphs length < 0xFFFF)

**Reply by behdad** [reply]
It's a hack that seems to work. I've seen other fonts do the same.

You shouldn't need sorting anymore.

@justvanrossum should we commit this?

**Reply by justvanrossum** [reply]
Not sure. Does ots accept it? Should the spec be adjusted? Does fonttools round trip this?

**Reply by behdad** [reply]
Good questions...

FWIW the kern table has similar issue and we already accept such fonts (calibri.ttf for example). Looks like ots-sanitize drops the kern table of it though, but claims success. @khaledhosny

**Comment by behdad** [comment]
In this code:
```
        format = 4 
        subTable_3_1 = buildCmapSubTable(cmapping_3_1, format, 3, 1) 
        try: 
            subTable_3_1.compile(self.font) 
        except struct.error: 
            # format 4 overflowed, fall back to format 12 
            if not allowFallback: 
                raise ValueError( 
                    "cmap format 4 subtable overflowed; sort glyph order by unicode to fix." 
                ) 
            format = 12 
            subTable_3_1 = buildCmapSubTable(cmapping_3_1, format, 3, 1)
```
I think the latter should have plat/enc of 3,10; not 3,1. No? That might be why Windows doesn't accept the font. @justvanrossum

**Reply by behdad** [reply]
And in the code following it, 0/3 should become 0/4 I think, in case a format=12 is being built:
```
        subTables.append(subTable_3_1) 
        subTable_0_3 = buildCmapSubTable(cmapping_3_1, format, 0, 3) 
        subTables.append(subTable_0_3) 
```

**Reply by TakWolf** [reply]
Will this issue be fixed? Is there a plan?

**Reply by behdad** [reply]
Can you share your font with us for testing?

**Reply by TakWolf** [reply]
Yes! What is needed?

Current problematic project is https://github.com/TakWolf/fusion-pixel-font

Recurrence the issues, run `build.py` on `master` branch, but need to set `pixel-font-builder==0.0.7` 

https://github.com/TakWolf/pixel-font-builder

What else do you need? I'll provide them!

**Reply by TakWolf** [reply]
This branch:  `issues/fonttools_discussions_3230`

https://github.com/TakWolf/fusion-pixel-font/tree/issues/fonttools_discussions_3230

**Reply by behdad** [reply]
Okay I have reproduced the result. If I make that change to fonttools to make it finish, which output file would be relevant, do you know?

**Reply by TakWolf** [reply]
If it can be fully built, all fonts will outputs to `build/outputs` dir,  otf and ttf.

Any font is ok, they just have different language flavors

**Reply by TakWolf** [reply]
This can just build 8px to accelerate the build speed.

```python

def main():
    fs_util.delete_dir(path_define.build_dir)

    update_service.setup_ark_pixel_glyphs()

    for font_config in configs.font_configs:
        # Just build 8px
        if font_config.size != 8:
            continue

```


Here can block out some unnecessary formats.

https://github.com/TakWolf/fusion-pixel-font/blob/master/services/font_service.py#L289


If there is any other issues, please let me know at any time.

**Reply by behdad** [reply]
Humm. When I ttx dump the length field of the cmap4 in those fonts, I don't see 65535...

**Reply by TakWolf** [reply]
Yes, currently there are about 48000 glyphs. Do you need more than 65535 glyphs? Or code points > 65535?

It can simulate this situation through a simple mapping.

**Reply by behdad** [reply]
No this should be fine. I'm just surprised that none of the cmap tables seems to be 65535, which is the number my patch would have put in it. I'll debug later.

**Reply by behdad** [reply]
Ah okay I think I need to look into the ttc...

**Comment by behdad** [comment]
Neither fonttools decompile nor ots-sanitize like this cmap.  Maybe we abort the idea.

@justvanrossum can you check my suggestion in https://github.com/fonttools/fonttools/discussions/3230#discussioncomment-6618953 ?  I can send a PR.

**Reply by TakWolf** [reply]
Does it mean that there is no way to solve this problem? 
I found that `min(length, 0xFFFF)` not really solving the problem. This just truncates the data. The generated data is still not compliant. (Although fonts can be compiled)

**Reply by justvanrossum** [reply]
The immediate way to likely solve the problem is to sort your glyph order by unicode, as has been suggested from the start.

Elswhere in this discussion, Behdad suggested some fixes for the on-overflow fallback mechanism in FontBuilder that may solve this without reordering the glyph order. But this requires implementing and testing those suggestions.

**Reply by TakWolf** [reply]
Sorting is possible for single fonts. But for OTC, there are some shared glyphs with multiple code points mapping. They can't determine the order. In different child font, they have different orders. This is the difficult point.

**Reply by justvanrossum** [reply]
It doesn't need to be completely sorted, but sorting _what you can_ may already make the difference. I'm not sure why it would matter that different child fonts would have different orders. Are you saing all fonts in the OTC have the same glyph order and should keep the same glyph order? If that's the case, you could try computing a single glyph order based on the cmap of one of them, and apply that new order to all, and see if it doesn't overflow.

**Comment by justvanrossum** [comment]
> can you check my suggestion in https://github.com/fonttools/fonttools/discussions/3230#discussioncomment-6618953 ? I can send a PR.

You may be correct there, if I read the [cmap recommendations](https://learn.microsoft.com/en-us/typography/opentype/spec/recom#cmap-table):

- If the font supports only characters in the Unicode Basic Multilingual Plane (U+0000 to U+FFFF): either platform 3, encoding 1; or platform 0, encoding 3. With either encoding, use a format 4 subtable.

- If the font supports any characters in a Unicode supplementary plane (U+10000 to U+10FFFF): either platform 3, encoding 10; or platform 0, encoding 4. With either encoding, use a format 12 subtable.

**Comment by TakWolf** [comment]
I'm sorry to waking up this thread again.

It has been a long time and I still cannot understand the core of this problem.

I have created a more simpler example:

```python
from fontTools.fontBuilder import FontBuilder
from fontTools.pens.t2CharStringPen import T2CharStringPen


def drawTestGlyph(pen: T2CharStringPen):
    pen.moveTo((100, 100))
    pen.lineTo((100, 1000))
    pen.curveTo((200, 900), (400, 900), (500, 1000))
    pen.lineTo((500, 100))
    pen.closePath()


def main():
    fb = FontBuilder(1024, isTTF=False)
    fb.setupGlyphOrder(['.notdef', 'A'])
    
    cmap = {}
    for code_point in range(0, 0xFFFF):
        cmap[code_point] = 'A'
    fb.setupCharacterMap(cmap)
    
    advanceWidths = {
        '.notdef': 600, 
        'A': 600,
    }

    familyName = 'HelloTestFont'
    styleName = 'TotallyNormal'
    version = '0.1'

    nameStrings = dict(
        familyName=dict(en=familyName, nl='HalloTestFont'),
        styleName=dict(en=styleName, nl='TotaalNormaal'),
        uniqueFontIdentifier='fontBuilder: ' + familyName + '.' + styleName,
        fullName=familyName + '-' + styleName,
        psName=familyName + '-' + styleName,
        version='Version ' + version,
    )

    pen = T2CharStringPen(600, None)
    drawTestGlyph(pen)
    charString = pen.getCharString()
    charStrings = {
        '.notdef': charString,
        'A': charString,
    }
    fb.setupCFF(nameStrings['psName'], {'FullName': nameStrings['psName']}, charStrings, {})
    lsb = {gn: cs.calcBounds(None)[0] for gn, cs in charStrings.items()}
    metrics = {}
    for gn, advanceWidth in advanceWidths.items():
        metrics[gn] = (advanceWidth, lsb[gn])
    fb.setupHorizontalMetrics(metrics)
    fb.setupHorizontalHeader(ascent=824, descent=200)
    fb.setupNameTable(nameStrings)
    fb.setupOS2(sTypoAscender=824, usWinAscent=824, usWinDescent=200)
    fb.setupPost()
    fb.save('test.otf')


if __name__ == '__main__':
    main()
```

This only create two glyphs. One is '.notdef', the other is 'A', and I mapping code point from 0 to 0xFFFF to 'A' by cmap.

This will cause `ValueError: cmap format 4 subtable overflowed; sort glyph order by unicode to fix.`

After testing, if cmap range out of `range(0, 0x7FEF)` will cause this error.

This means that the upper limit of cmap length is 32751. Is this true?

I tried to dump the SourceHan `SourceHanSansSC-VF.otf.woff2`, it's cmap contains 44853 mappings. How does it do this?

<img width="351" height="401" alt="image" src="https://github.com/user-attachments/assets/a4d6ef1f-ae78-42c8-9d92-e09eb5065259" />


----- Edit ---------

I also try dump `unifont-17.0.03.otf`, it's cmap has 58910 mappings. This includes the mapping of all BMP planes (removing invalid code points).

<img width="328" height="330" alt="image" src="https://github.com/user-attachments/assets/98e47d5a-6185-438a-a569-2f73fd714e74" />

**Reply by justvanrossum** [reply]
It seems to be a combination of number of glyphs referenced and the number of code points. See the `single_glyph` flag in this modified version of your script:

```python
from fontTools.fontBuilder import FontBuilder
from fontTools.pens.t2CharStringPen import T2CharStringPen


def drawTestGlyph(pen: T2CharStringPen):
    pen.moveTo((100, 100))
    pen.lineTo((100, 1000))
    pen.curveTo((200, 900), (400, 900), (500, 1000))
    pen.lineTo((500, 100))
    pen.closePath()


def main():
    single_glyph = False
    if single_glyph:
        N = 0x7FEF  # fails with 0x7FF0
    else:
        N = 0xFE79  # fails with 0xFE7A
    glyphOrder = ['.notdef'] + [f"glyph{i:05d}" for i in range(N)]
    fb = FontBuilder(1024, isTTF=False)
    fb.setupGlyphOrder(glyphOrder)
    
    cmap = {}
    for code_point, glyphName in enumerate(glyphOrder[1:]):
        cmap[code_point] = glyphOrder[0] if single_glyph else glyphName
    fb.setupCharacterMap(cmap)
    
    advanceWidths = {
        glyphName: 600
        for glyphName in glyphOrder
    }

    familyName = 'HelloTestFont'
    styleName = 'TotallyNormal'
    version = '0.1'

    nameStrings = dict(
        familyName=dict(en=familyName, nl='HalloTestFont'),
        styleName=dict(en=styleName, nl='TotaalNormaal'),
        uniqueFontIdentifier='fontBuilder: ' + familyName + '.' + styleName,
        fullName=familyName + '-' + styleName,
        psName=familyName + '-' + styleName,
        version='Version ' + version,
    )

    pen = T2CharStringPen(600, None)
    drawTestGlyph(pen)
    charString = pen.getCharString()
    charStrings = {
        glyphName: charString
        for glyphName in glyphOrder
    }
    fb.setupCFF(nameStrings['psName'], {'FullName': nameStrings['psName']}, charStrings, {})
    glyph_lsb = charString.calcBounds(None)[0]
    lsb = {gn: glyph_lsb for gn in charStrings.keys()}
    metrics = {}
    for gn, advanceWidth in advanceWidths.items():
        metrics[gn] = (advanceWidth, lsb[gn])
    fb.setupHorizontalMetrics(metrics)
    fb.setupHorizontalHeader(ascent=824, descent=200)
    fb.setupNameTable(nameStrings)
    fb.setupOS2(sTypoAscender=824, usWinAscent=824, usWinDescent=200)
    fb.setupPost()
    fb.save('test.otf')


if __name__ == '__main__':
    main()
```

**Reply by justvanrossum** [reply]
So, with, glyph IDs that go up with the code points, a cmap format 4 can encode 65145 (0xFE79) code points. 

FontBuilder's "sort glyph order by unicode to fix" error is misleading in this case. Either way, cmap format 4 is a very strange beast with hard-to-understand properties.
